example/src/main/java/io/netty/example/http2/tiles/http2/Http2RequestHandler.java

```diff
@@ -0,0 +1,138 @@
+
+        QueryStringDecoder queryString = new QueryStringDecoder(request.uri());
+        String streamId = streamId(request);
+
+        // Check arguments: path must match and latency parameter must be
+        // present
+        if (!PATH.equals(queryString.path()) || missing(queryString, LATENCY)) {
+            sendDummy(ctx, streamId);
+            return;
+        }
+
+        int latency = latency(queryString);
+
+        if (missing(queryString, X) && missing(queryString, Y)) {
+            handlePage(ctx, streamId, latency);
+        } else {
+            handleImage(queryString, ctx, streamId, latency);
+        }
+    }
+
+    private void handleImage(QueryStringDecoder query, ChannelHandlerContext ctx, String streamId, int latency) {
+        int x = toInt(query.parameters().get(X).get(0), 0);
+        int y = toInt(query.parameters().get(Y).get(0), 0);
+        byte[] image = ImageCache.instance().image(x, y);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, wrappedBuffer(image));
+        response.headers().set(CONTENT_TYPE, "image/jpeg");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    private void handlePage(ChannelHandlerContext ctx, String streamId, int latency) {
+        ByteBuf content = Unpooled.wrappedBuffer(Html.HEADER, Html.body(latency), Html.FOOTER);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, content);
+        response.headers().set(CONTENT_TYPE, "text/html; charset=UTF-8");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    protected void sendResponse(final ChannelHandlerContext ctx, String streamId, int latency,
+            final FullHttpResponse response) {
+        setContentLength(response, response.content().readableBytes());
+        setStreamId(response, streamId);
+        ctx.executor().schedule(new Runnable() {
+            @Override
+            public void run() {
+                ctx.writeAndFlush(response);
+            }
+        }, latency, TimeUnit.MILLISECONDS);
+    }
+
+    private int latency(QueryStringDecoder queryString) {
+        int latency = toInt(queryString.parameters().get(LATENCY).get(0), 0);
+        // Make sure that latency is not too big
+        return abs(latency > 1000 ? 0 : latency);
+    }
+
+    private String streamId(FullHttpRequest request) {
+        return request.headers().get(HttpUtil.ExtensionHeaderNames.STREAM_ID.text());
+    }
+
+    private void setStreamId(FullHttpResponse response, String streamId) {
+        response.headers().set(HttpUtil.ExtensionHeaderNames.STREAM_ID.text(), streamId);
+    }
+
+    private boolean missing(QueryStringDecoder query, String string) {
+        List<String> values = query.parameters().get(string);
+        return query.parameters() == null || values == null || values.size() == 0 || EMPTY_STRING.equals(values.get(0));
```

Comment:
Sorry I should have been more clear.  I'm not necessarily suggesting adding/using a new method in `QueryStringDecoder` but instead change the `private boolean missing(...)` method to the following:


```java
private static String firstValue(QueryStringDecoder query, String key) {
  if (query == null) { throw new NullPointerException("query"); }
  List<String> values = query.parameters().get(key);
  if (values == null || values.isEmpty() {
    return null;
  }
  return values.get(0);
}
```

Then using this method you can restructure things to avoid the double lookup you are currently doing against `query.parameters()` (i.e. https://github.com/netty/netty/pull/3620/files#diff-949d866cd8ab6ebed33601fde8c17581R80).